metrics/api/src/main/java/io/helidon/metrics/api/AbstractRegistry.java

```diff
@@ -0,0 +1,808 @@
+        if (metricIDs == null || metricIDs.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(
+                new AbstractMap.SimpleEntry<>(allMetrics.get(metricIDs.get(0)), metricIDs));
+    }
+
+    /**
+     * Returns a stream of {@link Map.Entry} for this registry.
+     *
+     * @return Stream of {@link Map.Entry}
+     */
+    protected Stream<Map.Entry<MetricID, M>> stream() {
+        return allMetrics.entrySet().stream();
+    }
+
+    /**
+     * Creates a new instance of an implementation wrapper around the indicated metric.
+     *
+     * @param metadata {@code Metadata} for the metric
+     * @param metric the existing metric to be wrapped by the impl
+     * @param <T> specific type of {@code Metric} provided and wrapped
+     * @return new wrapper implementation around the specified metric instance
+     */
+    protected abstract <T extends Metric> M toImpl(Metadata metadata, T metric);
+
+    protected abstract Map<MetricType, BiFunction<String, Metadata, M>> prepareMetricFactories();
+
+    // -- Package private -----------------------------------------------------
+
+    protected Optional<Map.Entry<MetricID, M>> getOptionalMetricEntry(String metricName) {
+        return getOptionalMetricWithIDsEntry(metricName).map(entry -> {
+                final MetricID metricID = entry.getValue().get(0);
+                return new AbstractMap.SimpleImmutableEntry<>(metricID,
+                        allMetrics.get(metricID));
+        });
+    }
+
+    private Optional<M> getOptionalMetric(String metricName, Tag... tags) {
+        return getOptionalMetric(new MetricID(metricName, tags));
+    }
+
+    private Optional<M> getOptionalMetric(MetricID metricID) {
+        return Optional.ofNullable(allMetrics.get(metricID));
+    }
+
+    protected List<Map.Entry<MetricID, M>> getMetricsByName(String metricName) {
+        List<MetricID> metricIDs = allMetricIDsByName.get(metricName);
+        if (metricIDs == null) {
+            return Collections.emptyList();
+        }
+        List<Map.Entry<MetricID, M>> result = new ArrayList<>();
+        for (MetricID metricID : metricIDs) {
+            result.add(new AbstractMap.SimpleEntry<>(metricID, allMetrics.get(metricID)));
+        }
+        return result;
+    }
+
+    protected Type registryType() {
+        return type;
+    }
+
+    protected List<MetricID> metricIDsForName(String metricName) {
+        return allMetricIDsByName.get(metricName);
+    }
+
+    static <T extends Metadata, U extends Metadata> boolean  metadataMatches(T a, U b) {
+        if (a == b) {
+            return true;
+        }
+        if (a == null || b == null) {
+            return false;
+        }
+        return a.getName().equals(b.getName())
+                && a.getTypeRaw().equals(b.getTypeRaw())
+                && (a.getDisplayName().equals(b.getDisplayName())
+                && Objects.equals(a.getDescription(), b.getDescription())
```

Comment:
there is a misleading end bracket or misleading formatting.

Reply:
fixed